### PR TITLE
Add SetupKey auto-groups property 

### PR DESCRIPTION
**Pull request by braginini**
Extend Management API to add support for setup keys auto-tagging feature that allows automatic groups assignment to peers that used the key.
This PR just adds the property to the setup key.

Works as the first feature MVP for this request https://github.com/netbirdio/netbird/issues/458#issue-1366523452

### CODE REVIEW

**Comment by braginini** [comment]
Relates to
https://github.com/netbirdio/dashboard/pull/75

File: management/server/http/setupkeys.go [hunk 1]
```diff
@@ -118,54 +89,108 @@ func (h *SetupKeys) HandleKey(w http.ResponseWriter, r *http.Request) {
...
+		log.Errorf("failed getting setup key %s under account %s %v", keyID, account.Id, err)
+		http.Redirect(w, r, "/", http.StatusInternalServerError)
 		return
-	default:
-		http.Error(w, "", http.StatusNotFound)
 	}
-}
 
-func (h *SetupKeys) GetKeys(w http.ResponseWriter, r *http.Request) {
+	writeSuccess(w, key)
```
Comment (mlsmaycon) [inline management/server/http/setupkeys.go:110]: we should call toResponseBody function before responding.

Comment (mlsmaycon) [inline management/server/http/setupkeys.go:110]: we should call toResponseBody function before responding.

File: management/server/http/setupkeys.go [hunk 2]
```diff
@@ -118,54 +89,108 @@ func (h *SetupKeys) HandleKey(w http.ResponseWriter, r *http.Request) {
...
+
+	newKey := &server.SetupKey{}
+	newKey.AutoGroups = req.AutoGroups
+	newKey.Revoked = req.Revoked
+	newKey.Name = req.Name
+	newKey.Id = keyID
 
-		err = json.NewEncoder(w).Encode(respBody)
+	newKey, err = h.accountManager.SaveSetupKey(account.Id, newKey)
+	if err != nil {
```
Comment (mlsmaycon) [inline management/server/http/setupkeys.go:154]: extra if err != nil

File: management/server/http/setupkeys.go [hunk 3]
```diff
@@ -118,54 +89,108 @@ func (h *SetupKeys) HandleKey(w http.ResponseWriter, r *http.Request) {
...
+		http.Redirect(w, r, "/", http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(200)
+	w.Header().Set("Content-Type", "application/json")
+
+	// should be initialized like that, otherwise json marshaller will return null
+	respBody := []*api.SetupKey{}
+	for _, key := range account.SetupKeys {
```
Comment (mlsmaycon) [inline management/server/http/setupkeys.go]: we can follow routes new convention and call a list keys function in the account manager.

File: management/server/http/setupkeys.go [hunk 4]
```diff
@@ -190,16 +215,27 @@ func toResponseBody(key *server.SetupKey) *api.SetupKey {
 	} else {
 		state = "valid"
 	}
+
+	// the UpdatedAt field was introduced later, so there might be that some keys have a Zero value (e.g, null in the store file)
+	var updatedAt time.Time
+	if key.UpdatedAt.IsZero() {
```
Comment (mlsmaycon) [inline management/server/http/setupkeys.go]: this check is already performed in the account management methods

File: management/server/setupkey.go [hunk 5]
```diff
@@ -83,24 +127,25 @@ func (key *SetupKey) IsOverUsed() bool {
 }
 
 // GenerateSetupKey generates a new setup key
-func GenerateSetupKey(name string, t SetupKeyType, validFor time.Duration) *SetupKey {
+func GenerateSetupKey(name string, t SetupKeyType, validFor time.Duration, autoGroups []string) *SetupKey {
```
Comment (mlsmaycon) [inline management/server/setupkey.go:130]: we could keep some of these setup key functions and methods private, what do you think?

File: management/server/http/setupkeys.go [hunk 6]
```diff
@@ -190,16 +215,27 @@ func toResponseBody(key *server.SetupKey) *api.SetupKey {
...
+		Key:        key.Key,
+		Name:       key.Name,
+		Expires:    key.ExpiresAt,
+		Type:       string(key.Type),
+		Valid:      key.IsValid(),
+		Revoked:    key.Revoked,
+		UsedTimes:  key.UsedTimes,
+		LastUsed:   key.LastUsed,
+		State:      state,
+		AutoGroups: key.AutoGroups,
```
Comment (mlsmaycon) [inline management/server/http/setupkeys.go:225]: should we consider returning []api.GroupMinimum?

**Review by mlsmaycon — CHANGES_REQUESTED** [review]

File: management/server/http/setupkeys.go (hunk 1 above)
Comment (braginini) [reply management/server/http/setupkeys.go:110]: writeSuccess function does toResponseBody

```
func writeSuccess(w http.ResponseWriter, key *server.SetupKey) {
	w.WriteHeader(200)
	w.Header().Set("Content-Type", "application/json")
	err := json.NewEncoder(w).Encode(toResponseBody(key))
	if err != nil {
		http.Error(w, "failed handling request", http.StatusInternalServerError)
		return
	}
}
```

Comment (braginini) [reply management/server/http/setupkeys.go:110]: writeSuccess function does toResponseBody

```
func writeSuccess(w http.ResponseWriter, key *server.SetupKey) {
	w.WriteHeader(200)
	w.Header().Set("Content-Type", "application/json")
	err := json.NewEncoder(w).Encode(toResponseBody(key))
	if err != nil {
		http.Error(w, "failed handling request", http.StatusInternalServerError)
		return
	}
}
```

File: management/server/http/setupkeys.go (hunk 2 above)
Comment (braginini) [reply management/server/http/setupkeys.go:154]: ok

File: management/server/http/setupkeys.go (hunk 3 above)
Comment (braginini) [reply management/server/http/setupkeys.go]: ok

File: management/server/http/setupkeys.go (hunk 4 above)
Comment (braginini) [reply management/server/http/setupkeys.go]: not really for the list of setup keys, but I will make sure it is only there

File: management/server/setupkey.go (hunk 5 above)
Comment (braginini) [reply management/server/setupkey.go:130]: Good idea, but it is used somewhere and won't be available anymore.
I think that we should leave for the major refactor of the account manager. There are many functions to remove as well.

**Review by mlsmaycon — APPROVED** [review]

**Pull request merged**